File: src/ffmpeg/dag/validate.py
from __future__ import annotations

from dataclasses import replace

from ..exceptions import FFMpegValueError
from ..streams.audio import AudioStream
from ..streams.video import VideoStream
from .context import DAGContext
from .nodes import FilterNode, InputNode
from .schema import Node, Stream


def remove_split(
    current_stream: Stream, mapping: dict[Stream, Stream] = None
) -> tuple[Stream, dict[Stream, Stream]]:
    """
    Rebuild the graph with the given mapping.

    Args:
        current_stream: The stream to rebuild the graph with.
        mapping: The mapping to rebuild the graph with.

    Returns:
        A tuple of the new node and the new mapping.
    """

    # remove all split nodes
    # add split nodes to the graph
    if mapping is None:
        mapping = {}

    if current_stream in mapping:
        return mapping[current_stream], mapping

    if not current_stream.node.inputs:
        mapping[current_stream] = current_stream
        return current_stream, mapping

    if isinstance(current_stream.node, FilterNode):
        # if the current node is a split node, we need to remove it
        if current_stream.node.name in ("split", "asplit"):
            new_stream, _mapping = remove_split(
                current_stream=current_stream.node.inputs[0], mapping=mapping
            )
            mapping[current_stream] = mapping[current_stream.node.inputs[0]]
            return mapping[current_stream.node.inputs[0]], mapping

    inputs = {}
    for idx, input_stream in sorted(
        enumerate(current_stream.node.inputs),
        key=lambda idx_stream: -len(idx_stream[1].node.upstream_nodes),
    ):
        new_stream, _mapping = remove_split(
            current_stream=input_stream, mapping=mapping
        )
        inputs[idx] = new_stream
        mapping |= _mapping

    new_node = replace(
        current_stream.node,
        inputs=tuple(
            stream for idx, stream in sorted(inputs.items(), key=lambda x: x[0])
        ),
    )
    new_stream = replace(current_stream, node=new_node)

    mapping[current_stream] = new_stream
    return new_stream, mapping


def add_split(
    current_stream: Stream,
    down_node: Node = None,
    down_index: int = None,
    context: DAGContext = None,
    mapping: dict[tuple[Stream, Node | None, int | None], Stream] = None,
) -> tuple[Stream, dict[tuple[Stream, Node | None, int | None], Stream]]:
    """
    Add split nodes to the graph.

    Args:
        current_stream: The stream to add split nodes to.
        down_node: The node use current_stream as input.
        down_index: The index of the input stream in down_node.
        context: The DAG context.
        mapping: The mapping to add split nodes to.

    Returns:
        A tuple of the new node and the new mapping.
    """

    if not context:
        context = DAGContext.build(current_stream.node)

    if mapping is None:
        mapping = {}

    if (current_stream, down_node, down_index) in mapping:
        return mapping[(current_stream, down_node, down_index)], mapping

    inputs = {}

    for idx, input_stream in sorted(
        enumerate(current_stream.node.inputs),
        key=lambda idx_stream: -len(idx_stream[1].node.upstream_nodes),
    ):
        new_stream, _mapping = add_split(
            current_stream=input_stream,
            down_node=current_stream.node,
            down_index=idx,
            mapping=mapping,
            context=context,
        )
        inputs[idx] = new_stream
        mapping |= _mapping

    new_node = replace(
        current_stream.node,
        inputs=tuple(
            stream for idx, stream in sorted(inputs.items(), key=lambda x: x[0])
        ),
    )
    new_stream = replace(current_stream, node=new_node)

    num = len(context.get_outgoing_nodes(current_stream))
    if num < 2:
        mapping[(current_stream, down_node, down_index)] = new_stream
        return new_stream, mapping

    if isinstance(current_stream.node, InputNode):
        for idx, (node, index) in enumerate(context.get_outgoing_nodes(current_stream)):
            # if the current node is InputNode, we don't need to split it
            mapping[(current_stream, node, index)] = new_stream
        return new_stream, mapping

    if isinstance(new_stream, VideoStream):
        split_node = new_stream.split(outputs=num)
        for idx, (node, index) in enumerate(context.get_outgoing_nodes(current_stream)):
            mapping[(current_stream, node, index)] = split_node.video(idx)
        return mapping[(current_stream, down_node, down_index)], mapping
    elif isinstance(new_stream, AudioStream):
        split_node = new_stream.asplit(outputs=num)
        for idx, (node, index) in enumerate(context.get_outgoing_nodes(current_stream)):
            mapping[(current_stream, node, index)] = split_node.audio(idx)
        return mapping[(current_stream, down_node, down_index)], mapping
    else:
        raise FFMpegValueError(f"unsupported stream type: {current_stream}")


def fix_graph(stream: Stream) -> Stream:
    """
    Fix the graph by removing and adding split nodes.

    Args:
        stream: The stream to fix.

    Returns:
        The fixed stream.

    Note:
        Fix the graph by resetting split nodes.
        This function is for internal use only.
    """

    stream, _ = remove_split(stream)
    stream, _ = add_split(stream)
    return stream


def validate(stream: Stream, auto_fix: bool = True) -> Stream:
    """
    Validate the given DAG. If auto_fix is True, the graph will be automatically fixed to follow ffmpeg's rule.

    Args:
        stream: The DAG to validate.
        auto_fix: Whether to automatically fix the graph.

    Returns:
        The validated DAG context.
    """
    if auto_fix:
        stream = fix_graph(stream)

    # NOTE: we don't want to modify the original node
    # validators: list[] = []

    # for validator in validators:
    #     context = validator(context)

    return stream

File: packages/ts-core/src/compile/validate.ts
/**
 * Graph validation and transformation for FFmpeg filter chains.
 *
 * Handles automatic insertion of split/asplit filters where streams
 * are reused as input to multiple filters.
 */

import { StreamType } from "../common/schema.js";
import { FilterableStream } from "../dag/baseStreams.js";
import {
  AudioStream,
  AVStream,
  FilterNode,
  GlobalNode,
  GlobalStream,
  InputNode,
  OutputNode,
  OutputStream,
  SubtitleStream,
  VideoStream,
} from "../dag/nodes.js";
import { Node, Stream } from "../dag/schema.js";
import { FFMpegValueError } from "../exceptions.js";
import type { KwargsValue } from "../utils/frozenRecord.js";
import { DAGContext } from "./context.js";

type StreamMapping = Map<Stream, Stream>;
type SplitKey = string;
type SplitMapping = Map<SplitKey, Stream>;

function splitKey(stream: Stream, node: Node | null, index: number | null): SplitKey {
  return `${stream.hex}:${node === null ? "null" : node.hex}:${index}`;
}

/** Rebuild a node with new inputs, preserving all other properties. */
function rebuildNodeWithInputs(node: Node, newInputs: readonly Stream[]): Node {
  if (node instanceof FilterNode) {
    return new FilterNode(
      node.name,
      newInputs as FilterableStream[],
      { ...node.kwargs },
      [...node.inputTypings],
      [...node.outputTypings],
    );
  }
  if (node instanceof InputNode) {
    return node;
  }
  if (node instanceof OutputNode) {
    return new OutputNode(
      newInputs as FilterableStream[],
      node.filename,
      { ...node.kwargs },
    );
  }
  if (node instanceof GlobalNode) {
    return new GlobalNode(
      newInputs as OutputStream[],
      { ...node.kwargs },
    );
  }
  throw new FFMpegValueError(`Unknown node type: ${node.constructor.name}`);
}

/** Rebuild a stream pointing to a new node, preserving index and optional. */
function rebuildStream(stream: Stream, newNode: Node): Stream {
  if (stream instanceof VideoStream) {
    return new VideoStream(newNode, stream.index, stream.optional);
  }
  if (stream instanceof AudioStream) {
    return new AudioStream(newNode, stream.index, stream.optional);
  }
  if (stream instanceof AVStream) {
    return new AVStream(newNode, stream.index, stream.optional);
  }
  if (stream instanceof SubtitleStream) {
    return new SubtitleStream(newNode, stream.index, stream.optional);
  }
  if (stream instanceof OutputStream) {
    return new OutputStream(newNode as OutputNode);
  }
  if (stream instanceof GlobalStream) {
    return new GlobalStream(newNode as GlobalNode);
  }
  throw new FFMpegValueError(`Unknown stream type: ${stream.constructor.name}`);
}

/** Remove all split/asplit nodes from the graph. */
export function removeSplit(
  currentStream: Stream,
  mapping?: StreamMapping,
): [Stream, StreamMapping] {
  if (!mapping) mapping = new Map();
  if (mapping.has(currentStream)) return [mapping.get(currentStream)!, mapping];

  if (currentStream.node.inputs.length === 0) {
    mapping.set(currentStream, currentStream);
    return [currentStream, mapping];
  }

  if (currentStream.node instanceof FilterNode) {
    const name = currentStream.node.name;
    if (name === "split" || name === "asplit") {
      removeSplit(currentStream.node.inputs[0], mapping);
      mapping.set(currentStream, mapping.get(currentStream.node.inputs[0])!);
      return [mapping.get(currentStream.node.inputs[0])!, mapping];
    }
  }

  const indexedInputs = currentStream.node.inputs.map((s, i) => [i, s] as const);
  const sorted = [...indexedInputs].sort(
    (a, b) => b[1].node.upstreamNodes.size - a[1].node.upstreamNodes.size,
  );

  const newInputs = new Map<number, Stream>();
  for (const [idx, inputStream] of sorted) {
    const [newStream] = removeSplit(inputStream, mapping);
    newInputs.set(idx, newStream);
  }

  const orderedInputs = [...newInputs.entries()]
    .sort((a, b) => a[0] - b[0])
    .map(([, s]) => s);

  const newNode = rebuildNodeWithInputs(currentStream.node, orderedInputs);
  const newStream = rebuildStream(currentStream, newNode);

  mapping.set(currentStream, newStream);
  return [newStream, mapping];
}

/** Add split/asplit nodes where streams are reused. */
export function addSplit(
  currentStream: Stream,
  downNode: Node | null = null,
  downIndex: number | null = null,
  context?: DAGContext,
  mapping?: SplitMapping,
): [Stream, SplitMapping] {
  if (!context) context = DAGContext.build(currentStream.node);
  if (!mapping) mapping = new Map();

  const key = splitKey(currentStream, downNode, downIndex);
  if (mapping.has(key)) return [mapping.get(key)!, mapping];

  const indexedInputs = currentStream.node.inputs.map((s, i) => [i, s] as const);
  const sorted = [...indexedInputs].sort(
    (a, b) => b[1].node.upstreamNodes.size - a[1].node.upstreamNodes.size,
  );

  const newInputs = new Map<number, Stream>();
  for (const [idx, inputStream] of sorted) {
    const [newStream, newMapping] = addSplit(
      inputStream,
      currentStream.node,
      idx,
      context,
      mapping,
    );
    newInputs.set(idx, newStream);
    mapping = newMapping;
  }

  const orderedInputs = [...newInputs.entries()]
    .sort((a, b) => a[0] - b[0])
    .map(([, s]) => s);

  const newNode = rebuildNodeWithInputs(currentStream.node, orderedInputs);
  const newStream = rebuildStream(currentStream, newNode);

  const outgoing = context.getOutgoingNodes(currentStream);
  if (outgoing.length < 2) {
    mapping.set(key, newStream);
    return [newStream, mapping];
  }

  if (currentStream.node instanceof InputNode) {
    for (const [node, index] of outgoing) {
      mapping.set(splitKey(currentStream, node, index), newStream);
    }
    return [newStream, mapping];
  }

  if (newStream instanceof VideoStream) {
    const splitNode = new FilterNode(
      "split",
      [newStream],
      { outputs: outgoing.length },
      [StreamType.Video],
      Array(outgoing.length).fill(StreamType.Video) as StreamType[],
    );
    for (let idx = 0; idx < outgoing.length; idx++) {
      const [node, index] = outgoing[idx];
      mapping.set(splitKey(currentStream, node, index), splitNode.video(idx));
    }
    return [mapping.get(key)!, mapping];
  } else if (newStream instanceof AudioStream) {
    const splitNode = new FilterNode(
      "asplit",
      [newStream],
      { outputs: outgoing.length },
      [StreamType.Audio],
      Array(outgoing.length).fill(StreamType.Audio) as StreamType[],
    );
    for (let idx = 0; idx < outgoing.length; idx++) {
      const [node, index] = outgoing[idx];
      mapping.set(splitKey(currentStream, node, index), splitNode.audio(idx));
    }
    return [mapping.get(key)!, mapping];
  }

  throw new FFMpegValueError(
    `Unsupported stream type for splitting: ${currentStream.constructor.name}`,
  );
}

/** Fix stream reuse issues by removing then re-adding split nodes. */
export function fixGraph(stream: Stream): Stream {
  const [withoutSplits] = removeSplit(stream);
  const [withSplits] = addSplit(withoutSplits);
  return withSplits;
}

/** Validate a filter graph and optionally fix stream reuse issues. */
export function validate(stream: Stream, autoFix: boolean = true): Stream {
  if (autoFix) {
    return fixGraph(stream);
  }
  return stream;
}
